java-frontend/src/main/java/org/sonar/java/se/checks/SwitchCaseWithoutBreakCheck.java

```diff
@@ -0,0 +1,134 @@
+import org.sonar.check.Rule;
+import org.sonar.java.se.CheckerContext;
+import org.sonar.java.se.ProgramState;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.CaseLabelTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.SwitchStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+
+/**
+ * Rule implementation which checks for switch-case fall-through.
+ */
+@Rule(key = "S128")
+public class SwitchCaseWithoutBreakCheck extends SECheck {
+
+  private static final String ISSUE_MESSAGE
+    = "End this switch case with an unconditional break, return or throw statement.";
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private static boolean isUnconditionalExit(Tree syntaxNode) {
+    if (syntaxNode.is(
+      Tree.Kind.BREAK_STATEMENT,
+      Tree.Kind.THROW_STATEMENT,
+      Tree.Kind.RETURN_STATEMENT,
+      // The continue statement also unconditionally exits the switch-statement.
+      Tree.Kind.CONTINUE_STATEMENT)) {
+      // Exit found
+      return true;
+    }
+
+    if (!syntaxNode.is(Tree.Kind.BLOCK, Tree.Kind.IF_STATEMENT, Tree.Kind.TRY_STATEMENT)) {
+      // Not a block of code or a condition, not a possible exit.
+      return false;
+    }
+
+    if (syntaxNode.is(Tree.Kind.BLOCK)) {
+      // Tests if the block contains an unconditional statement.
+      BlockTree block = (BlockTree) syntaxNode;
+      return block.body().stream().anyMatch(SwitchCaseWithoutBreakCheck::isUnconditionalExit);
+    } else if (syntaxNode.is(Tree.Kind.TRY_STATEMENT)) {
+      return isUnconditionalExitInTryCatchConstruct((TryStatementTree) syntaxNode);
+    } else {
+      return isUnconditionalExitInIfStatement((IfStatementTree) syntaxNode);
+    }
+  }
+
+  private static boolean isUnconditionalExitInTryCatchConstruct(TryStatementTree tryStatement) {
+    if (!isUnconditionalExit(tryStatement.block())) {
+      return false;
+    }
+
+    // If the try-block exits unconditionally then all catch blocks must do so as well.
+    // When they don't the try-catch block is not a valid unconditional exit.
+    return tryStatement.catches().stream().allMatch(catchTree -> isUnconditionalExit(catchTree.block()));
+  }
+
+  private static boolean isUnconditionalExitInIfStatement(IfStatementTree ifStatement) {
+    if (!isUnconditionalExitInStatement(ifStatement.thenStatement())) {
+      return false;
+    }
+
+    StatementTree elseStatement = ifStatement.elseStatement();
+    if (elseStatement == null) {
+      // Without else this is a conditional exit.
+      return false;
+    }
+
+    return isUnconditionalExitInStatement(elseStatement);
```

Comment:
You can probably call directly isUnconditionalExit(elseStatement) here rather than using a dedicated one for StatementTree.